### PR TITLE
Ivy-popupEditorBackup-controller-unit-inte-test

### PR DESCRIPTION
**Pull request by 20chen-7**
# Description
Please include the exact bug/functionality description and a summary of the changes/ related issues. Please also include any other relevant motivation and context:
Fixes # (bug list priority high/medium/low x.y.z)
Or Implements # (WBS) 

## Related PRS (if any):
This frontend PR is related to the #XXX backend PR.
To test this backend PR you need to checkout the #XXX frontend PR.
…

## Main changes explained:
- Add createPopupEditorBackup.md
- Add getAllPopupEditorBackups.md
- Add getPopupEditorBackupById.md
- Add updatePopupEditorBackup.md
- Add popupEditorBackupController.spec.js
- Add popupEditorBackupRouter.test.js
…

## How to test:
1. check into current branch
2. do `npm install` and `...` to run this PR locally
3. do `npm run test popupEditorBackup`

## Screenshots or videos of changes:
Screenshot 2024-06-06 at 11 34 30 AM

## Note:
Include the information the reviewers need to know.

### CODE REVIEW

**Review by Kavil-Jain-514 — APPROVED** [review]
All test cases have passed without any error or warning.

image

**Review by Logeshwari-Renu — APPROVED** [review]
All the test cases have passed.! 

PR976

**Review by deepthikannan — APPROVED** [review]
image
All tests for popupEditorBackup have passed successfully, indicating that the implementation is correct and complete – approved!

File: src/controllers/popupEditorBackupController.js [hunk 1]
```diff
@@ -1,10 +1,10 @@
...
 const popupEditorBackupController = function (PopupEditorBackups) {
   const getAllPopupEditorBackups = function (req, res) {
     PopupEditorBackups.find()
-      .then(results => res.status(200).send(results))
-      .catch(error => res.status(404).send(error));
+      .then((results) => res.status(200).send(results))
+      .catch((error) => res.status(404).send(error));
   };
 
   const getPopupEditorBackupById = function (req, res) {
```
Comment (DiegoSalas27) [inline src/controllers/popupEditorBackupController.js:10]: The piece of code down here is incorrect:
```
try {
      PopupEditorBackups.find({ popupId: { $in: popupId } }, (error, popupBackup) => {
        res.status(200).send(popupBackup[0]);
      });
    } catch (error) {
      res.status(404).send(error);
    }
```
We pass a callback function as a second parameter, and while that is allowed, the error is not being accounted for. The exception you throw in your unit test actually is a programming error of doing: `null[0]`. This is wrong, and should be fixed. Replace with the following code instead:

```
try {
      const popupBackup = await PopupEditorBackups.findById(popupId);
      res.status(200).send(popupBackup);
    } catch (error) {
      res.status(404).send(error);
    }
```

make sure you update `const getPopupEditorBackupById = function (req, res) ` to   const `getPopupEditorBackupById = async function (req, res)` since we are using async await here.

The intention is to getPopupEditorBackup by id, so even using `find` is nonsensical. `FindById` is what you should use. More: https://mongoosejs.com/docs/api/model.html#Model.findById(). `find` actually returns an empty array if no documents match the query, and as a consequence `[][0]` returns undefined. The frontend should expect either a model or undefined.

Update test cases accordingly.

File: src/controllers/popupEditorBackupController.js [hunk 2]
```diff
@@ -64,15 +61,16 @@ const popupEditorBackupController = function (PopupEditorBackups) {
       PopupEditorBackups.find({ popupId: { $in: popupId } }, (error, popupBackup) => {
```
Comment (DiegoSalas27) [inline src/controllers/popupEditorBackupController.js:61]: Same, here. We passs a callback function, but the error of `find` will get lost. The `catch` block will only catch errors that happen inside of the callback function but not the `error` parameter from the callback function if `find` method where to error out. Also, we are updating a single entry of `PopupEditorBackups`.

There should be something like findOneAndUpdate() or findByIdAndUpdate() or you can just do findById() then update the entity and save it back again. This refactor is optional, since the case where `popupBackp` is an empty array is accounted for. But still,  the code isn't clean. and I cannot imagine what could happen if an actual error is thrown by `find` method. What would `popupBackup` look like? If it's anything different than an array, we would throw a programming error.

Something to bring up to the team if necessary.

**Review by DiegoSalas27 — DISMISSED** [review]
Great job with your tests. I can definitely see the effort and the thoroughness of your tests. You accounted for edge cases and even tested edge cases in your integration tests. Very well done. Left some comments, please address them and I will approve this PR.

**Comment by akshit0211** [comment]
Hi, all test cases passed. Good job!

all test cases passed 1
all test cases passed 2

**Review by akshit0211 — APPROVED** [review]

**Review by Ramaruva — APPROVED** [review]
All test cases are passed good work!
image

**Review by aaryaneil — APPROVED** [review]
Tested the PR and all the test cases passed.
Screenshot 2024-06-15 at 4 59 14 AM

**Review by jinxyou — APPROVED** [review]
All test cases have passed. Great work! 
Screenshot 2024-06-15 151620

**Review by shradac — APPROVED** [review]
All tests passed. The tests cover all specified scenarios. Good Work!

**Comment by mianmiantea2019** [comment]
All tests passed. Good Work! 

image

**Comment by sahithidhulipala9** [comment]
All test cases passed!
PR 976

**Review by sahithidhulipala9 — APPROVED** [review]

**Review by Rishitha1309 — APPROVED** [review]
All test cases passed successfully.
PR976

**Review by nikhilpittala16 — APPROVED** [review]
I have tested the PR and all test cases have passed.
Screenshot (70)

**Review by gowthamdongari — APPROVED** [review]
All test cases passed without any issues
image

**Review by Annezhang673 — APPROVED** [review]
All tests passed. 
PR976

**Review by SatyashanthiT — APPROVED** [review]
I have tested this PR by checking out the current branch, installing dependencies, and running the tests locally. All tests for createPopupEditorBackup, getAllPopupEditorBackups, getPopupEditorBackupById, and updatePopupEditorBackup passed successfully. 

image

**Comment by Gmon-k** [comment]
All the test cases has passed and Below is screenshot for it
